tmt/utils/__init__.py

```diff
@@ -5892,22 +5896,95 @@ def is_key_origin(node: fmf.Tree, key: str) -> bool:
     return origin is not None and node.name == origin.name
 
 
-def resource_files(path: Union[str, Path], package: Union[str, ModuleType] = "tmt") -> Path:
+# TODO: Move this in a dedicated module
+@functools.cache
+def _get_resource_files_search_path(
+    package: Union[str, ModuleType], logger: tmt.log.Logger
+) -> Union[Path, MultiplexedPath]:
     """
-    Helper function to get path of package file or directory.
+    Helper (cached) function for :py:func:`resource_files`.
+
+    :param package: primary package in which to search for the file/directory.
+    :param logger: logger to report plugin import failures
+    :returns: the search path for resource files
+    """
+
+    package_path = MultiplexedPath(importlib.resources.files(package))
+
+    # Additional resource files can be imported from entry-point
+    entry_point_name = 'tmt.resources'
+
+    entry_point_group = importlib.metadata.entry_points(group=entry_point_name)
+
+    final_paths = [package_path]
+    for ep in entry_point_group:
+        try:
+            ep_module = ep.load()
+            ep_path = MultiplexedPath(importlib.resources.files(ep_module))
+            final_paths.append(ep_path)
+        except ModuleNotFoundError:
+            logger.warning(f"Failed to load plugin resources: {ep}")
+        except Exception as err:
+            # Other exceptions are rather weird
+            logger.warning(f"Unexpected failure in parsing: {ep}\n{err}")
+    return MultiplexedPath(*final_paths)
+
+
+@overload
+def resource_files(
+    path: Union[str, Path],
+    package: Union[str, ModuleType] = "tmt",
+    *,
+    logger: tmt.log.Logger,
+    assert_file: Literal[True],
+) -> Path: ...
 
-    A thin wrapper for :py:func:`importlib.resources.files`:
-    ``files()`` returns ``Traversable`` object, though in our use-case
-    it should always produce a :py:class:`pathlib.PosixPath` object.
-    Converting it to :py:class:`tmt.utils.Path` instance should be
-    safe and stick to the "``Path`` only!" rule in tmt's code base.
 
-    :param path: file or directory path to retrieve, relative to the ``package`` root.
-    :param package: package in which to search for the file/directory.
-    :returns: an absolute path to the requested file or directory.
+@overload
+def resource_files(
+    path: Union[str, Path],
+    package: Union[str, ModuleType] = "tmt",
+    *,
+    logger: tmt.log.Logger,
+    assert_file: Literal[False] = False,
+) -> Union[Path, MultiplexedPath]: ...
+
+
+def resource_files(
+    path: Union[str, Path],
+    package: Union[str, ModuleType] = "tmt",
+    *,
+    logger: tmt.log.Logger,
+    assert_file: bool = False,
```

Comment:
1, `assert_file` is undocumented, 2. it feels ad hoc-ish, easy to omit. Would it be possible to provide `resource_file()` and `resource_directory()`, for example? Then it would be clear to a developer and code what they are after and what they expect.